pkg/controller/atlasstream/atlasstream_connection.go

```diff
@@ -0,0 +1,256 @@
+	return func(streamConnection *akov2.AtlasStreamConnection) (*admin.StreamsConnection, error) {
+		connection := admin.StreamsConnection{
+			Name: &streamConnection.Spec.Name,
+			Type: &streamConnection.Spec.ConnectionType,
+		}
+
+		if streamConnection.Spec.ClusterConfig != nil {
+			connection.ClusterName = &streamConnection.Spec.ClusterConfig.Name
+			connection.DbRoleToExecute = &admin.DBRoleToExecute{
+				Role: &streamConnection.Spec.ClusterConfig.Role.Name,
+				Type: &streamConnection.Spec.ClusterConfig.Role.RoleType,
+			}
+		}
+
+		if streamConnection.Spec.KafkaConfig != nil {
+			authData, err := getSecretData(
+				ctx,
+				k8sClient,
+				*streamConnection.Spec.KafkaConfig.Authentication.Credentials.GetObject(streamConnection.Namespace),
+				kafkaConnectionAuthUsername,
+				kafkaConnectionAuthPassword,
+			)
+			if err != nil {
+				return nil, err
+			}
+
+			secData, err := getSecretData(
+				ctx,
+				k8sClient,
+				*streamConnection.Spec.KafkaConfig.Security.Certificate.GetObject(streamConnection.Namespace),
+				kafkaConnectionSecCertificate,
+			)
+			if err != nil {
+				return nil, err
+			}
+
+			connection.BootstrapServers = &streamConnection.Spec.KafkaConfig.BootstrapServers
+			connection.Authentication = &admin.StreamsKafkaAuthentication{
+				Mechanism: &streamConnection.Spec.KafkaConfig.Authentication.Mechanism,
+				Username:  pointer.MakePtr(authData[kafkaConnectionAuthUsername]),
+				Password:  pointer.MakePtr(authData[kafkaConnectionAuthPassword]),
+			}
+			connection.Security = &admin.StreamsKafkaSecurity{
+				BrokerPublicCertificate: pointer.MakePtr(secData[kafkaConnectionSecCertificate]),
+				Protocol:                &streamConnection.Spec.KafkaConfig.Security.Protocol,
+			}
+			connection.Config = &streamConnection.Spec.KafkaConfig.Config
+		}
+
+		return &connection, nil
+	}
+}
+
+func getSecretData(ctx context.Context, k8sClient client.Client, ref client.ObjectKey, keys ...string) (map[string]string, error) {
+	secret := corev1.Secret{}
+	err := k8sClient.Get(ctx, ref, &secret)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve secret %v: %w", ref, err)
+	}
+
+	data := map[string]string{}
+	for _, key := range keys {
+		val, ok := secret.Data[key]
+		if !ok {
+			return nil, fmt.Errorf("key %s is not present in the secret %v", key, ref)
+		}
+
+		data[key] = string(val)
+	}
+
+	return data, nil
+}
+
+func hasStreamConnectionChanged(
+	streamConnection *akov2.AtlasStreamConnection,
+	atlasStreamConnection admin.StreamsConnection,
+	mapper streamConnectionMapper,
+) (bool, error) {
+	// Unset API metadata for comparison
+	atlasStreamConnection.Links = nil
+	connection, err := mapper(streamConnection)
+	if err != nil {
+		return false, err
+	}
+	if _, ok := connection.GetAuthenticationOk(); ok {
+		connection.Authentication.Password = nil
+	}
+
+	return !reflect.DeepEqual(*connection, atlasStreamConnection), nil
```

Comment:
We cannot do this, this will fail on nested slices in the `admin.StreamsConnection` type, i.e. `.links`, `.authentication.links`, and `.security.links`. We have two options:

1. clear (i.e. set to `nil`) any non-comparable fields in `atlasStreamConnection`

or

2. use normalization, where:
- we convert `atlasStreamConnection` to `*akov2.AtlasStreamConnection`
- use normalization from https://github.com/mongodb/mongodb-atlas-kubernetes/tree/semanticequals
- `reflect.DeepEqual` our own `*akov2.AtlasStreamConnection` types

Option 2. is preferred but not a showstopper for this PR.

In any case this is brittle as it relies on non-slices in embedded fields in `admin.StreamsConnection` so we have to make it very clear in the godoc of this function that we rely on the invariant of no slices being embedded in this type.